File: core/integration/testdata/modules/java/defaults/src/main/java/io/dagger/modules/defaults/Defaults.java
package io.dagger.modules.defaults;

import io.dagger.client.DaggerQueryException;
import io.dagger.client.Directory;
import io.dagger.client.File;
import io.dagger.module.AbstractModule;
import io.dagger.module.annotation.Default;
import io.dagger.module.annotation.DefaultPath;
import io.dagger.module.annotation.Function;
import io.dagger.module.annotation.Ignore;
import io.dagger.module.annotation.Object;
import java.util.concurrent.ExecutionException;
import java.util.Optional;

@Object
public class Defaults extends AbstractModule {
  public Defaults() {
    super();
  }

  @Function
  public String echo(@Default("default value") String value) {
    return value;
  }

  @Function
  public String echoElse(Optional<String> value) {
    return value.orElse("default value if null");
  }

  @Function
  public String echoOptDefault(@Default("default value") Optional<String> value) {
    return value.get();
  }

  @Function
  public String fileName(@DefaultPath("dagger.json") File file)
      throws ExecutionException, DaggerQueryException, InterruptedException {
    return file.name();
  }

  @Function
  public String fileNames(@DefaultPath("src/main/java/io/dagger/modules/defaults") Directory dir)
      throws ExecutionException, DaggerQueryException, InterruptedException {
    return String.join(" ", dir.entries());
  }

  @Function
  public String filesNoIgnore(@DefaultPath(".") Directory dir)
      throws ExecutionException, DaggerQueryException, InterruptedException {
    return String.join(" ", dir.entries());
  }

  @Function
  public String filesIgnore(@DefaultPath(".") @Ignore({"*.xml"}) Directory dir)
      throws ExecutionException, DaggerQueryException, InterruptedException {
    return String.join(" ", dir.entries());
  }

  @Function
  public String filesNegIgnore(@DefaultPath(".") @Ignore({"**", "!**/*.java"}) Directory dir)
      throws ExecutionException, DaggerQueryException, InterruptedException {
    return String.join(" ", dir.entries());
  }
}

File: core/integration/testdata/modules/elixir/defaults/lib/defaults.ex
defmodule Defaults do
  @moduledoc false

  use Dagger.Mod.Object, name: "Defaults"

  defn echo_else(value: String.t() | nil) :: String.t() do
    if(value, do: value, else: "default value if null")
  end

  defn file_name(file: {Dagger.File.t(), default_path: "dagger.json"}) :: String.t() do
    Dagger.File.name(file)
  end

  defn file_names(dir: {Dagger.Directory.t(), default_path: "lib"}) :: String.t() do
    with {:ok, entries} <- Dagger.Directory.entries(dir) do
      Enum.join(entries, " ")
    end
  end

  defn files_no_ignore(dir: {Dagger.Directory.t(), default_path: "."}) :: String.t() do
    with {:ok, entries} <- Dagger.Directory.entries(dir) do
      Enum.join(entries, " ")
    end
  end

  defn files_ignore(dir: {Dagger.Directory.t(), default_path: ".", ignore: ["mix.exs"]}) :: String.t() do
    with {:ok, entries} <- Dagger.Directory.entries(dir) do
      Enum.join(entries, " ")
    end
  end

  defn files_neg_ignore(dir: {Dagger.Directory.t(), default_path: ".", ignore: ["**", "!**/*.ex"]}) :: String.t() do
    with {:ok, entries} <- Dagger.Directory.entries(dir) do
      Enum.join(entries, " ")
    end
  end
end
